README.md

# workboots <img src="man/figures/logo.png" align="right" width="120" />

**Author:** [Mark Rieke](https://www.thedatadiary.net/about/) <br/>
**License:**
[MIT](https://github.com/markjrieke/workboots/blob/main/LICENSE)

<!-- badges: start -->

[![R-CMD-check](https://github.com/markjrieke/workboots/workflows/R-CMD-check/badge.svg)](https://github.com/markjrieke/workboots/actions)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/workboots)](https://CRAN.R-project.org/package=workboots)
[![](https://cranlogs.r-pkg.org/badges/grand-total/workboots)](https://cran.r-project.org/package=workboots)
<!-- badges: end -->

## Overview

`{workboots}` is a tidy method of generating bootstrap prediction
intervals for arbitrary model types from a tidymodel workflow.

By using [bootstrap
resampling](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)),
we can create many models — one for each resample. Each model will be
slightly different based on the resample it was trained on. Each model
will also generate slightly different predictions for new data, allowing
us to generate a prediction distribution for models that otherwise just
return point predictions.

## Installation

You can install the released version of workboots from CRAN or the
development version from github with the
[devtools](https://cran.r-project.org/package=devtools) or
[remotes](https://cran.r-project.org/package=remotes) package:

``` r
# install from CRAN
install.packages("workboots")

# or the development version
devtools::install_github("markjrieke/workboots")
```

## Usage

workboots builds on top of the `{tidymodels}` suite of packages.
Teaching how to use tidymodels is beyond the scope of this package, but
some helpful resources are linked at the bottom of this README.

To get started, we’ll need to create a workflow.

``` r
library(tidymodels)

# load our dataset
data("penguins")
penguins <- penguins %>% drop_na()

# split data into testing & training sets
set.seed(123)
penguins_split <- initial_split(penguins)
penguins_test <- testing(penguins_split)
penguins_train <- training(penguins_split)

# create a workflow
penguins_wf <- 
  workflow() %>%
  add_recipe(recipe(body_mass_g ~ ., data = penguins_train) %>% step_dummy(all_nominal())) %>%
  add_model(boost_tree("regression"))
```

Boosted tree models can only generate point predictions, but with
workboots we can generate a prediction interval for each observation in
`penguins_test` by passing the workflow to `predict_boots()`:

``` r
library(workboots)

# generate predictions from 2000 bootstrap models
set.seed(345)
penguins_pred_int <-
  penguins_wf %>%
  predict_boots(
    n = 2000,
    training_data = penguins_train,
    new_data = penguins_test
  )

# summarise predictions with a 95% prediction interval
pengins_pred_int %>%
  summarise_predictions()
```

    #> # A tibble: 84 x 5
    #>    rowid .preds               .pred .pred_lower .pred_upper
    #>    <int> <list>               <dbl>       <dbl>       <dbl>
    #>  1     1 <tibble [2,000 x 2]> 3465.       2913.       3994.
    #>  2     2 <tibble [2,000 x 2]> 3535.       2982.       4100.
    #>  3     3 <tibble [2,000 x 2]> 3604.       3050.       4187.
    #>  4     4 <tibble [2,000 x 2]> 4157.       3477.       4764.
    #>  5     5 <tibble [2,000 x 2]> 3868.       3305.       4372.
    #>  6     6 <tibble [2,000 x 2]> 3519.       2996.       4078.
    #>  7     7 <tibble [2,000 x 2]> 3435.       2914.       3954.
    #>  8     8 <tibble [2,000 x 2]> 4072.       3483.       4653.
    #>  9     9 <tibble [2,000 x 2]> 3445.       2926.       3966.
    #> 10    10 <tibble [2,000 x 2]> 3405.       2876.       3938.
    #> # ... with 74 more rows

Alternatively, we can generate a confidence interval around each
prediction by setting the parameter `interval` to `"confidence"`:

``` r
# generate predictions from 2000 bootstrap models
set.seed(456)
penguins_conf_int <- 
  penguins_wf %>%
  predict_boots(
    n = 2000,
    training_data = penguins_train,
    new_data = penguins_test,
    interval = "confidence"
  )

# summarise with a 95% confidence interval
penguins_conf_int %>%
  summarise_predictions()
```

    #> # A tibble: 84 x 5
    #>    rowid .preds               .pred .pred_lower .pred_upper
    #>    <int> <list>               <dbl>       <dbl>       <dbl>
    #>  1     1 <tibble [2,000 x 2]> 3466.       3257.       3635.
    #>  2     2 <tibble [2,000 x 2]> 3534.       3291.       3811.
    #>  3     3 <tibble [2,000 x 2]> 3623.       3306.       3921.
    #>  4     4 <tibble [2,000 x 2]> 4155.       3722.       4504.
    #>  5     5 <tibble [2,000 x 2]> 3868.       3644.       4086.
    #>  6     6 <tibble [2,000 x 2]> 3509.       3286.       3768.
    #>  7     7 <tibble [2,000 x 2]> 3439.       3249.       3624.
    #>  8     8 <tibble [2,000 x 2]> 4064.       3737.       4369.
    #>  9     9 <tibble [2,000 x 2]> 3450.       3253.       3635.
    #> 10    10 <tibble [2,000 x 2]> 3405.       3222.       3651.
    #> # ... with 74 more rows

## Bug reports/feature requests

If you notice a bug, want to request a new feature, or have
recommendations on improving documentation, please [open an
issue](https://github.com/markjrieke/workboots/issues) in this
repository.

### Tidymodels Resources

-   [Getting started with Tidymodels](https://www.tidymodels.org/start/)
-   [Tidy Modeling with R](https://www.tmwr.org/)
-   [Julia Silge’s Blog](https://juliasilge.com/blog/) provides use
    cases of tidymodels with weekly
    [\#tidytuesday](https://github.com/rfordatascience/tidytuesday)
    datasets.


NEWS.md
# workboots 0.2.0

### Function updates

* Rearranged column order output of `summarise_*` functions from `*_lower`, `*`, `*_upper` to `*`, `*_lower`, `*_upper`
* Deprecated `conf` parameter in `summarise_*` functions in favor of `interval_width`.
* Added support for generating confidence intervals from `predict_boots()`. 

# workboots 0.1.1

### Function updates

* Updates to `predict_boots()`
  + updated function to generate prediction interval (previously was generating a prediction's confidence interval)
  + updated default setting to assume residuals are normally distributed
  + updated default number of resamples
  + updated function to draw residuals based on the [632+ rule](https://stats.stackexchange.com/questions/96739/what-is-the-632-rule-in-bootstrapping) (previously was using training residuals)
* updated default number of resamples in `vi_boots()`
* added param `verbose` to both `predict_boots()` and `vi_boots()` to display progress in the console. 
* added new function `summarise_importance()`, as well alias `summarize_*` for `summarise_*`

### Bug fixes

* Fixed bug in `assert_pred_data()` that caused some workflows to be rejected by `predict_boots()`

# workboots 0.1.0

* Initial release
* Core functions:
  + `predict_boots()` for generating bootstrap prediction intervals from tidymodel workflows.
  + `summarise_predictions()` for summarizing bootstrap predictions with expected, lower bound, and upper bound values.
  + `vi_boots()` for generating bootstrap feature importance from tidymodel workflows.
* Vignette for Getting Started with workboots.
* Unit testing for core functions (via testthat).


R/predict_boots.R
#' Fit and predict from a workflow using many bootstrap resamples.
#'
#' Generate a prediction interval from arbitrary model types using bootstrap
#' resampling. `predict_boots()` generates `n` bootstrap resamples, fits a model
#' to each resample (creating `n` models), then creates `n` predictions for each
#' observation in `new_data`.
#'
#' @details Since `predict_boots()` fits a new model to each resample, the
#'  argument `workflow` must not yet be fit. Any tuned hyperparameters must be
#'  finalized prior to calling `predict_boots()`.
#'
#' @return A tibble with a column indicating the row index of each observation in
#'  `new_data` and a nested list of the model predictions for each observation.
#'
#' @param new_data A tibble or dataframe used to make predictions.
#' @param interval One of `prediction`, `confidence`. Specifies the interval type to be generated.
#' @inheritParams vi_boots
#'
#' @export
#'
#' @importFrom rlang arg_match
#' @importFrom rlang warn
#' @importFrom rsample bootstraps
#' @importFrom purrr map_dfc
#' @importFrom tibble rowid_to_column
#' @importFrom tidyr pivot_longer
#' @importFrom tidyr nest
#'
#' @examples
#' \dontrun{
#' library(tidymodels)
#'
#' # setup a workflow without fitting
#' wf <-
#'   workflow() %>%
#'   add_recipe(recipe(qsec ~ wt, data = mtcars)) %>%
#'   add_model(linear_reg())
#'
#' # fit and predict 2000 bootstrap resampled models to mtcars
#' set.seed(123)
#' wf %>%
#'   predict_boots(n = 2000, training_data = mtcars, new_data = mtcars)
#' }
predict_boots <- function(workflow,
                          n = 2000,
                          training_data,
                          new_data,
                          interval = c("prediction", "confidence"),
                          verbose = FALSE,
                          ...) {

  # convert interval type
  interval <- rlang::arg_match(interval)

  # check arguments
  assert_workflow(workflow)
  assert_n(n)
  assert_pred_data(workflow, training_data, "training")
  assert_pred_data(workflow, new_data, "new")

  # warn if low n
  if (n < 2000) {

    rlang::warn(
      paste0("At least 2000 resamples recommended for stable results.")
    )

  }

  # create resamples from training set
  training_boots <-
    rsample::bootstraps(
      training_data,
      times = n,
      ...
    )

  # map sequence of indices to `predict_single_boot()`
  # returns a column of predictions for each model
  preds <-
    purrr::map_dfc(
      seq(1, n),
      ~predict_single_boot(
        workflow = workflow,
        boot_splits = training_boots,
        new_data = new_data,
        interval = interval,
        verbose = verbose,
        index = .x
      )
    )

  # nest & return predictions in long format
  preds <- tibble::rowid_to_column(preds)

  preds <-
    tidyr::pivot_longer(
      preds,
      dplyr::starts_with(".pred_"),
      names_to = "model",
      values_to = "model.pred"
    )

  preds <-
    tidyr::nest(
      preds,
      .preds = c(model, model.pred)
    )

  return(preds)

}

# --------------------------------internals-------------------------------------

#' (Internal) Generate a column of predictions on new data based on a model fit
#' to a single training bootstrap.
#'
#' @param workflow passed from `predict_boots()`
#' @param boot_splits passed from `predict_boots()`
#' @param new_data passed from `predict_boots()`
#' @param verbose passed from `predict_boots()`
#' @param index passed from `predict_boots()`
#'
#' @importFrom rsample training
#' @importFrom rsample testing
#' @importFrom generics fit
#' @importFrom dplyr filter
#' @importFrom dplyr pull
#' @importFrom stats predict
#' @importFrom rlang sym
#' @importFrom tidyr crossing
#' @importFrom Metrics rmse
#' @importFrom stats sd
#' @importFrom tibble add_column
#' @importFrom stats rnorm
#' @importFrom dplyr mutate
#' @importFrom dplyr rename
#' @importFrom rlang :=
#'
#' @noRd
#'
predict_single_boot <- function(workflow,
                                boot_splits,
                                new_data,
                                interval,
                                verbose,
                                index) {

  # get training data from bootstrap resample split
  boot_train <-
    rsample::training(
      boot_splits$splits[[index]]
    )

  # get oob sample
  boot_oob <-
    rsample::testing(
      boot_splits$splits[[index]]
    )

  # fit workflow to training data
  model <- generics::fit(workflow, boot_train)

  # predict given model and new data
  preds <- stats::predict(model, new_data)

  # get predicted var name
  pred_name <- dplyr::filter(workflow$pre$actions$recipe$recipe$var_info, role == "outcome")
  pred_name <- dplyr::pull(pred_name, variable)

  # apply prediction interval using bootstrap 632+ estimate
  # if not, just returns absolute prediction (when summarised, this generates a confidence interval)
  if (interval == "prediction") {

    # get training residuals
    preds_train <- dplyr::pull(stats::predict(model, boot_train), .pred)
    actuals_train <- dplyr::pull(boot_train, rlang::sym(pred_name))
    resids_train <- actuals_train - preds_train
    resids_train <- resids_train - mean(resids_train)

    # get oob residuals
    preds_oob <- dplyr::pull(stats::predict(model, boot_oob), .pred)
    actuals_oob <- dplyr::pull(boot_oob, rlang::sym(pred_name))
    resids_oob <- actuals_oob - preds_oob
    resids_oob <- resids_oob - mean(resids_oob)

    # calculate no-information error rate (rmse_ni) with RMSE as loss function
    combos <- tidyr::crossing(actuals_train, preds_train)
    rmse_ni <- Metrics::rmse(combos$actuals_train, combos$preds_train)

    # calculate overfit rate
    rmse_oob <- Metrics::rmse(actuals_oob, preds_oob)
    rmse_train <- Metrics::rmse(actuals_train, preds_train)
    overfit <- (rmse_oob - rmse_train)/(rmse_ni - rmse_train)

    # calculate weight (if overfit = 0, weight = .632 & residual used will just be .632)
    # uses the actual proportion of distinct training/oob samples, rather than average of 0.632/0.368
    prop_368 <- nrow(boot_oob)/nrow(boot_train)
    prop_632 <- 1 - prop_368
    weight <- prop_632/(1 - (prop_368 * overfit))

    # determine residual std.dev based on weight
    sd_oob <- stats::sd(resids_oob)
    sd_train <- stats::sd(resids_train)
    sd_resid <- weight * sd_oob + (1 - weight) * sd_train

    # add residuals to fit
    preds <- tibble::add_column(preds, resid_add = stats::rnorm(nrow(new_data), 0, sd_resid))
    preds <- dplyr::mutate(preds, .pred = .pred + resid_add)
    preds <- preds[, 1]

  }

  # rename .pred col based on index number
  preds <- dplyr::rename(preds, !!rlang::sym(paste0(".pred_", index)) := .pred)

  # print progress when verbose is set to TRUE
  verbose_print(verbose, index, nrow(boot_splits))

  return(preds)

}


